File: Camp Progress/547. Number of Provinces.cpp
// Question Link: https://leetcode.com/problems/number-of-provinces/

class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        vector<int> rank(n, 1);
        vector<int> dsu(n);
        for (int i = 0; i < n; i++) {
            dsu[i] = i;
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (isConnected[i][j] == 1) {
                    Union(i, j, dsu, rank);
                    isConnected[i][j] = 0;
                    isConnected[j][i] = 0;
                }
            }   
        }
        
        unordered_set<int> provinces;
        for (int i = 0; i < n; i++) {
            provinces.insert(find(dsu[i], dsu));
        }
        
        return provinces.size();
    }
    
    int find(int cur, vector<int> &dsu) {
        if (dsu[cur] == cur) {
            return cur;
        }
        dsu[cur] = find(dsu[cur], dsu);
        return dsu[cur];
    }
    
    void Union(int a, int b, vector<int> &dsu, vector<int> &rank) {
        int A = find(a, dsu);
        int B = find(b, dsu);
        
        if (A == B) {
            return;
        }
        
        if (rank[B] > rank[A]) {
            swap(B, A);
        }
        dsu[B] = A;
        
        rank[A] += rank[B];
    }
};

File: Python3 Solutions/547. Number of Provinces.py
# Question Link: https://leetcode.com/problems/number-of-provinces/

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        dsu = [ _ for _ in range(n) ]
        rank = [1]*n
        
        for i in range(n):
            for j in range(n):
                if isConnected[i][j] == 1:
                    self.union(i, j, dsu, rank)
                    isConnected[i][j] = 0
                    isConnected[j][i] = 0
            
        provinces = set()
        for p in dsu:
            provinces.add(self.find(dsu[p], dsu))
                
        return len(provinces)

    def find(self, c, dsu):
        if dsu[c] == c:
            return c
        dsu[c] = self.find(dsu[c], dsu)
        return dsu[c]
    
    def union(self, a, b, dsu, rank):
        A = self.find(a, dsu)
        B = self.find(b, dsu)
        
        if dsu[A] == dsu[B]:
            return
        
        dsu[B] = A
        
        if rank[B] > rank[A]:
            A, B = B, A
        
        rank[A] += rank[B]
        